src/box/lua/connpool.lua

```diff
@@ -75,21 +75,45 @@ local function connect_to_candidates(candidates)
     local delay = WATCHER_DELAY
     local conn_opts = {connect_timeout = WATCHER_TIMEOUT}
     local connected_candidates = {}
+
+    local connection_fibers = {}
+
     for _, instance_name in pairs(candidates) do
-        local time_connect_end = clock.monotonic() + WATCHER_TIMEOUT
         local ok, conn = pcall(connect, instance_name, conn_opts)
```

Comment:
Hm. `net_box.connect()` plus `:wait_connected()` still called synchronously. This code parallelize the RO/RW mode waiting, but a network problem likely affect connecting too.

Reply:
Thank you. Fixed. Now, it uses an approach with `on_connect` hook.

Reply:
I like the direction of your idea. However, it should be implemented in a different way. Flaws in the current implementation:

1. Connections with `mode == nil` now added into the list of connected candidates.
2. `:wait_connected()` in a loop may wait too long and unblock the fiber after the deadline.

I propose to add a fiber cond to signal about the `mode ~= nil` event from `connect` and spin in a loop of `cond:wait(remaining_time)`. The exit conditions for the loop are either:

1. The time quota is exceeded.
2. All the waiting connections have `mode ~= nil`.

At least, it works in my head this evening. Need to check, of course :)

Reply:
Yep. You're absolutely right. It should be implemented in a different way. Applied the discussed way.

Now there is a conditional variable that is triggered when some of connection modes are changed. We start connecting and wait for the moment when all of the instances are available until the timeout.

Actually, the perfect implementation would be like this (Lua-like pseudocode):
```lua
-- Creating connections:
function create_conn()
  -- ...
  conn = establish_conn()
  conn:on_update_mode(function() cond:broadcast() end)
  -- Currently the implementation lacks this.
  conn:on_error(function cond:broadcast() end)
  -- ...
end
-- Connecting to a candidates:
function connect_to_candidates()
  connect_to_all()
  while not the timeout reached do
    -- We may wait more than currently specified small delay.
    cond:wait(timeout - clock())
  end
end
```

The problem is in the absence of `on_error()` or `on_access()` connection triggers. Probably some of them should be implemented in order to fire the conditional variable when there is an error.

Reply:
`:wait_state()` is in the [public API](https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/#conn-wait-state) and it returns on reaching a terminal state.

https://github.com/tarantool/tarantool/blob/90552e55e0921405c43ea086ae418a72c9f000e4/src/box/lua/net_box.lua#L1025-L1035

Not exactly what you're asking for, but may be useful.

Reply:
Thank you. It seems useful, but not for that case. Because even using this the checks of the error will still be made once in some quite a small time interval. Is the current implementation considered ok?

Reply:
Sorry for the long delay! The implementation looks OK.

Maybe it worth to re-collect connected instances after exiting the loop by the timeout. OTOH, that's nigligible.